File: mc/rmcp/src/main/java/calebxzhou/rdi/mc/common2/mcp/RQuest.java
package calebxzhou.rdi.mc.common2.mcp;

import java.util.List;

/**
 * calebxzhou @ 2026-05-10 11:23
 */
public record RQuest(
        String id,
        String title,
        String subtitle,
        String chapterId,
        String chapterTitle,
        Position position,
        State state,
        Rules rules,
        List<String> description,
        List<Dependency> dependencies,
        List<Task> tasks,
        List<Reward> rewards
) {
    public record Position(double x, double y) {
    }

    public record State(
            boolean visible,
            boolean started,
            boolean completed,
            boolean startable,
            int progress,
            String cannotStartReason
    ) {
    }

    public record Rules(
            String progressionMode,
            String dependencyRequirement,
            int minRequiredDependencies,
            boolean optional,
            boolean repeatable,
            boolean hideDetailsUntilStartable,
            boolean requireSequentialTasks
    ) {
    }

    public record Dependency(
            String id,
            String type,
            String title,
            boolean visible,
            boolean started,
            boolean completed
    ) {
    }

    public record Task(
            String id,
            String type,
            String title,
            boolean completed,
            long progress,
            long maxProgress
    ) {
    }

    public record Reward(
            String id,
            String type,
            String title,
            boolean claimed
    ) {
    }
}

File: mc/rmcp/common/src/main/kotlin/calebxzhou/rdi/mc/common2/mcp/RQuest.kt
package calebxzhou.rdi.mc.common2.mcp

/**
 * calebxzhou @ 2026-05-10 11:23
 */
@JvmRecord
data class RQuest(
    val id: String,
    val title: String,
    val subtitle: String,
    val chapterId: String,
    val chapterTitle: String,
    val groupId: String,
    val groupTitle: String,
    val position: Position,
    val state: State,
    val rules: Rules,
    val description: List<String>,
    val dependencies: List<Dependency>,
    val tasks: List<Task>,
    val rewards: List<Reward>
) {
    @JvmRecord
    data class Position(val x: Double, val y: Double)

    @JvmRecord
    data class State(
        val visible: Boolean,
        val started: Boolean,
        val completed: Boolean,
        val startable: Boolean,
        val progress: Int,
        val cannotStartReason: String?
    )

    @JvmRecord
    data class Rules(
        val progressionMode: String,
        val dependencyRequirement: String,
        val minRequiredDependencies: Int,
        val optional: Boolean,
        val repeatable: Boolean,
        val hideDetailsUntilStartable: Boolean,
        val requireSequentialTasks: Boolean
    )

    @JvmRecord
    data class Dependency(
        val id: String,
        val type: String,
        val title: String,
        val visible: Boolean,
        val started: Boolean,
        val completed: Boolean
    )

    @JvmRecord
    data class Task(
        val id: String,
        val type: String,
        val title: String,
        val completed: Boolean,
        val progress: Long,
        val maxProgress: Long
    )

    @JvmRecord
    data class Reward(
        val id: String,
        val type: String,
        val title: String,
        val claimed: Boolean
    )
}
